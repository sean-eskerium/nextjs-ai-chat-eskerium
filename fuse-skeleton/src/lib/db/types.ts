/**
 * Database schema types that match our actual database tables
 */

export interface DbUser {
    id: string;
    email: string;
    role: string[];
    display_name: string | null;
    photo_url: string;
    password: string | null;
    shortcuts: string[];
    settings: Record<string, any>;
    login_redirect_url: string;
    email_verified: Date | null;
}

/**
 * Type for creating a new user in the database
 */
export type CreateDbUser = Omit<DbUser, 'id' | 'email_verified'> & {
    id?: string;
    email_verified?: Date | null;
}; 