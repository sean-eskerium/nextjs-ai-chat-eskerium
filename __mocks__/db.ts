export const deleteChatById = jest.fn();
export const getChatById = jest.fn();
export const getChats = jest.fn();
export const createChat = jest.fn();
export const updateChat = jest.fn(); 