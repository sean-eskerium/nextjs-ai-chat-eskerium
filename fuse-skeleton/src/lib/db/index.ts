import { drizzle } from 'drizzle-orm/postgres-js';
import postgres from 'postgres';

const connectionString = process.env.POSTGRES_PRISMA_URL!;

// For migrations
export const migrationClient = postgres(connectionString, { max: 1 });

// For queries
const queryClient = postgres(connectionString);
export const db = drizzle(queryClient); 