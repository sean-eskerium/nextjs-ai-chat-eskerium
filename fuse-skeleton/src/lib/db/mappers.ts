import type { User, DbUser } from '@auth/types';

/**
 * Convert a database user to a UI user model
 */
export function dbToUser(dbUser: DbUser): User {
    return {
        id: dbUser.id,
        email: dbUser.email,
        role: dbUser.role,
        name: dbUser.name || '',
        displayName: dbUser.displayName || '',
        photoURL: dbUser.photoURL || '',
        image: dbUser.image || '',
        password: dbUser.password,
        shortcuts: dbUser.data?.shortcuts || [],
        settings: dbUser.data?.settings || {},
        emailVerified: dbUser.emailVerified
    };
}

/**
 * Convert a UI user model to database fields
 */
export function userToDb(user: Partial<User> & { password?: string }): Partial<DbUser> {
    const data = {
        shortcuts: user.shortcuts || [],
        settings: user.settings || {}
    };

    return {
        ...(user.id && { id: user.id }),
        ...(user.email && { email: user.email }),
        ...(user.role && { role: Array.isArray(user.role) ? user.role : [user.role] }),
        ...(user.name && { name: user.name }),
        ...(user.displayName && { displayName: user.displayName }),
        ...(user.photoURL && { photoURL: user.photoURL }),
        ...(user.image && { image: user.image }),
        ...(user.password && { password: user.password }),
        ...(user.emailVerified && { emailVerified: user.emailVerified }),
        data
    };
} 