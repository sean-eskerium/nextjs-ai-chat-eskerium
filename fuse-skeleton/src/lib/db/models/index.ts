export * from './user';
// We can add other models as needed 