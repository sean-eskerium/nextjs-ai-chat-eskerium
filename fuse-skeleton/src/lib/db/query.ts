'use server';

import { drizzle } from 'drizzle-orm/postgres-js';
import postgres from 'postgres';
import * as schema from './models';

// Declare but don't initialize outside the function
let queryClient: ReturnType<typeof postgres> | null = null;
let db: ReturnType<typeof drizzle<typeof schema>> | null = null;

// Create a function to get the db instance
export async function getDb() {
    if (typeof window !== 'undefined') {
        throw new Error('Database can only be accessed on the server side');
    }

    if (!queryClient) {
        console.log('getDb - Initializing database connection');
        queryClient = postgres(process.env.DATABASE_URL!, {
            max: 1,
            ssl: 'require',
            connect_timeout: 10,
        });

        console.log('getDb - Initializing drizzle with schema');
        db = drizzle(queryClient, { schema });
    }

    console.log('getDb - Returning db instance');

    return db;
}

// Export a type for use in other files
export type DB = NonNullable<ReturnType<typeof getDb>>; 