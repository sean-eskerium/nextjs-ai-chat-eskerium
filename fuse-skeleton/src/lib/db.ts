'use server';

import { getDb, type DB } from './db/query';

// Helper function to ensure we're on the server
export async function ensureServerSide() {
    if (typeof window !== 'undefined') {
        throw new Error('This function can only be called on the server side');
    }
}

export { getDb, type DB }; 