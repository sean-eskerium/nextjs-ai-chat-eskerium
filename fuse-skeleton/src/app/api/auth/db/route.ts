import { NextRequest, NextResponse } from 'next/server';
import { db, ensureServerSide } from '@/lib/db';
import { eq } from 'drizzle-orm';
import { user } from '@/lib/db/schema';

export async function GET(request: NextRequest) {
    // Ensure we're on the server side
    ensureServerSide();

    const email = request.nextUrl.searchParams.get('email');
    if (!email) {
        return NextResponse.json({ error: 'Email is required' }, { status: 400 });
    }

    try {
        if (!db) {
            throw new Error('Database connection not available');
        }

        const result = await db.query.user.findFirst({
            where: (u) => eq(u.email, email)
        });

        if (!result) {
            return NextResponse.json({ error: 'User not found' }, { status: 404 });
        }

        return NextResponse.json(result);
    } catch (error) {
        console.error('Database error:', error);
        return NextResponse.json({ error: 'Internal server error' }, { status: 500 });
    }
}

export async function PUT(request: NextRequest) {
    // Ensure we're on the server side
    ensureServerSide();

    try {
        const userData = await request.json();
        const email = userData.email;
        if (!email) {
            return NextResponse.json({ error: 'Email is required' }, { status: 400 });
        }

        if (!db) {
            throw new Error('Database connection not available');
        }

        const result = await db.query.user.findFirst({
            where: (u) => eq(u.email, email)
        });

        if (!result) {
            return NextResponse.json({ error: 'User not found' }, { status: 404 });
        }

        // Only update fields that are defined in the schema
        const updateData = {
            name: userData.name !== undefined ? userData.name : result.name,
            role: userData.role !== undefined ? userData.role : result.role,
            emailVerified: userData.emailVerified !== undefined 
                ? userData.emailVerified ? new Date(userData.emailVerified) : null 
                : result.emailVerified
        };

        const updatedUser = await db
            .update(user)
            .set(updateData)
            .where(eq(user.email, email))
            .returning();

        return NextResponse.json(updatedUser[0]);
    } catch (error) {
        console.error('Database error:', error);
        return NextResponse.json({ error: 'Internal server error' }, { status: 500 });
    }
} 