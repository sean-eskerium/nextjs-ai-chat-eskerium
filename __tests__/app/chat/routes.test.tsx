import { auth } from '@/app/(auth)/auth';
import { getChatById, getMessagesByChatId } from '@/lib/db/queries';
import { notFound } from 'next/navigation';
import { render } from '@testing-library/react';
import Page from '@/app/(chat)/chat/[id]/page';

// Mock nanoid
jest.mock('nanoid', () => ({
  nanoid: () => 'test-id',
}));

// Mock auth
jest.mock('@/app/(auth)/auth', () => ({
  auth: jest.fn(() => Promise.resolve({ user: { id: 'test-user' } })),
}));

// Mock database queries
jest.mock('@/lib/db/queries', () => ({
  getChatById: jest.fn(),
  getMessagesByChatId: jest.fn(() => Promise.resolve([])),
}));

// Mock next/navigation
jest.mock('next/navigation', () => ({
  notFound: jest.fn(),
}));

// Mock components
jest.mock('@/components/chat', () => ({
  Chat: () => <div data-testid="chat-component" />,
}));

jest.mock('@/components/data-stream-handler', () => ({
  DataStreamHandler: () => <div data-testid="data-stream-handler" />,
}));

describe('Chat Routes', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('Chat ID Route', () => {
    it('loads existing chat', async () => {
      const mockChat = {
        id: 'test-chat',
        title: 'Test Chat',
        visibility: 'public',
        userId: 'test-user',
      };

      (getChatById as jest.Mock).mockResolvedValueOnce(mockChat);
      (getMessagesByChatId as jest.Mock).mockResolvedValueOnce([]);

      const page = await Page({ params: Promise.resolve({ id: 'test-chat' }) });
      const { container } = render(page);

      expect(container).toMatchSnapshot();
      expect(getChatById).toHaveBeenCalledWith({ id: 'test-chat' });
      expect(notFound).not.toHaveBeenCalled();
    });

    it('returns not found for non-existent chat', async () => {
      (getChatById as jest.Mock).mockResolvedValueOnce(null);
      (getMessagesByChatId as jest.Mock).mockResolvedValueOnce([]);

      await Page({ params: Promise.resolve({ id: 'non-existent' }) });

      expect(getChatById).toHaveBeenCalledWith({ id: 'non-existent' });
      expect(notFound).toHaveBeenCalled();
    });

    it('returns not found for private chat without auth', async () => {
      const mockChat = {
        id: 'private-chat',
        title: 'Private Chat',
        visibility: 'private',
        userId: 'other-user',
      };

      (getChatById as jest.Mock).mockResolvedValueOnce(mockChat);
      (getMessagesByChatId as jest.Mock).mockResolvedValueOnce([]);
      (auth as jest.Mock).mockResolvedValueOnce(null);

      await Page({ params: Promise.resolve({ id: 'private-chat' }) });

      expect(getChatById).toHaveBeenCalledWith({ id: 'private-chat' });
      expect(notFound).toHaveBeenCalled();
    });
  });
}); 