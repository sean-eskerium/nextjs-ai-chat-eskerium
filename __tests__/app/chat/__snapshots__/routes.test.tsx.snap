// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat Routes Chat ID Route loads existing chat 1`] = `
<div>
  <div
    data-testid="chat-component"
  />
  <div
    data-testid="data-stream-handler"
  />
</div>
`;
